### PR TITLE
tunnel: various improvements

### PR DESCRIPTION
**Pull request by fatih**
* tunnel: various small improvements and tunings 
* client: add `StartNotify()` method to let know consumers a successful start
* client: improve and add new tests

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -262,18 +287,21 @@ func (c *Client) listenControl(ct *control) error {
...
-		c.log.Debug("controlMsg: %+v", msg)
+		c.log.Debug("Received control msg %+v", msg)
 
 		switch msg.Action {
 		case requestClientSession:
+			c.log.Debug("Received request to open a session to server")
 			go func() {
 				if err := c.proxy(msg.LocalPort); err != nil {
-					fmt.Fprintf(os.Stderr, "proxy err: '%s'\n", err)
+					c.log.Error("Proxy err between remote and local: '%s'\n", err)
```
Comment (cihangir) [inline client.go]: we shouldn't need new lines for log messages, logging package handles them

File: server.go [hunk 2]
```diff
@@ -149,12 +150,18 @@ func (s *Server) handleHTTP(w http.ResponseWriter, r *http.Request) error {
 
 	// ask client to open a session to us, so we can accept it
 	if err := control.send(msg); err != nil {
-		return err
+		// we might have several issues here, either the stream is closed, or
+		// the session is going be shut down, the underlying connection might
+		// be broken. In all cases, it's not reliable anymore having a client
+		// session.
+		control.Close()
```
Comment (cihangir) [inline server.go]: after closing the control connection, better to remove it  too `s.deleteControl`

File: server.go [hunk 3]
```diff
@@ -215,7 +230,8 @@ func (s *Server) controlHandler(w http.ResponseWriter, r *http.Request) (ctErr e
 		return fmt.Errorf("control conn for %s already exist. \n", identifier)
```
Comment (cihangir) [inline server.go]: another deletion of control conn is required here

**Comment by fatih** [comment]
Merging this. Thanks @cihangir for the review.

**Pull request merged**